benchmark/rust-ffi/app/Main.hs

```diff
@@ -0,0 +1,85 @@
+module Main where
+
+import Control.DeepSeq (force)
+import Control.Exception (evaluate)
+import Control.Monad (void, when)
+import Data.Foldable (for_, traverse_)
+import Data.Word (Word8)
+import Foreign.C.Types (CSize (..))
+import Foreign.Marshal (with, withArray)
+import Foreign.Ptr (Ptr, castPtr)
+import Foreign.Storable (Storable (sizeOf), alignment, peek, poke, pokeByteOff, sizeOf)
+import GHC.Compact (compact, getCompact)
+import System.IO (hPrint, stderr, stdout)
+import Text.Printf (hPrintf)
+
+import qualified Data.Primitive.ByteArray as BA
+import qualified System.Clock as Clock
+import qualified System.Environment as Env
+
+import Data.Text.Utf8 (CodeUnitIndex (..), Text (..))
+
+import qualified Data.Text.Utf8 as Utf8
+
+foreign import ccall unsafe "perform_ac"
+  performAc :: CSize -> Ptr U8Slice -> Ptr U8Slice -> IO CSize
+
+data U8Slice = U8Slice (Ptr Word8) CSize CSize
+
+instance Storable U8Slice where
+  sizeOf _ = sizeOf (undefined :: Ptr Word8) + 2 * sizeOf (undefined :: CSize)
+  alignment _ = max (alignment (undefined :: Ptr Word8)) (alignment (undefined :: CSize))
+  peek _ptr = error "We only write U8Slices to pointer, never read them"
+  poke ptr (U8Slice u8ptr off len) = do
+    poke ptr' u8ptr
+    pokeByteOff ptr' (sizeOf (undefined :: Ptr Word8)) off
+    pokeByteOff ptr' (sizeOf (undefined :: Ptr Word8) + sizeOf (undefined :: CSize)) len
+    where ptr' = castPtr ptr
+
+fromText :: Text -> U8Slice
+fromText (Text u8data off len) = U8Slice (BA.byteArrayContents u8data) (fromIntegral off) (fromIntegral len)
```

Comment:
Isn't this highly unsafe because the text bytearrays are not pinned?

It is safe to pass a non-pinned `ByteArray#` directly to an `unsafe` foreign import, as no GC can run during an unsafe FFI call, but with the indirection of first getting the pointer and building more things, and then doing the FFI call, you have a window for a GC to happen.

Reply:
Ah, I missed the compact part. Never mind then.
To make that more explicit, you could add a guard like this:

```haskell
 fromText :: Text -> U8Slice 
  | BA.isByteArrayPinned u8data = ... -- never mind then
  | otherwise = error "ByteArray must be pinned"
```

Then it's safe and clear why.

The pinning check does take into account compact data:
https://gitlab.haskell.org/ghc/ghc/-/blob/master/rts/PrimOps.cmm#L148-159